commands/bootstrap

```diff
@@ -297,6 +332,331 @@ create_self_signed() {
+
+    # Make the signing certificate trusted by the device
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" cp "$LOCAL_SIGNING_CERT_FILE" "/usr/local/share/ca-certificates"
+    "${EXEC_CMD[@]}" "$TARGET" "$SUDO" update-ca-certificates
+}
+
+# Create a certificate that is trusted by all the entities attached to a main device
+#
+# This certificate is created on the target device,
+# but signed on the host running the `c8y tedge` command on behalf of the main device.
+create_local_cert() {
+    MAIN_DEVICE_ID=$1
+    SERVICE=$2
+    if [ -z "$MAIN_DEVICE_ID" ]; then
+        echo "mTLS cannot be enabled without an id for the main device" >&2
+        exit 1
+    fi
+    if [ -z "$SERVICE" ]; then
+        SERVICE=main-agent
+    fi
+
+    # Prepare required info for subjectAltName
+    TARGET_DNS=$("${EXEC_CMD[@]}" "$TARGET" hostname 2>/dev/null || true)
+    if [ -z "$TARGET_DNS" ]; then
+        TARGET_DNS=$(echo "$TARGET" | sed -E 's/.*@(\S+)/\1/')
+    fi
+    TARGET_IP=$("${EXEC_CMD[@]}" "$TARGET" ip route get 8.8.8.8 2>/dev/null | sed -E 's/.*src (\S+) .*/\1/;t;d' 2>/dev/null || true)
+    if [ -z "$TARGET_IP" ]; then
+        TARGET_IP=$(echo "$TARGET" | sed -E 's/.*@(\S+)/\1/')
+    fi
+
+    # Prepare cert files
+    SIGNING_CERT_DIR="${HOME}/.tedge/device-signing-certs"
+    SIGNING_CERT_NAME="${MAIN_DEVICE_ID}-local-ca"
+    SIGNING_CERT_FILE="${SIGNING_CERT_DIR}/${SIGNING_CERT_NAME}.crt"
+    SIGNING_CERT_KEY="${SIGNING_CERT_DIR}/${SIGNING_CERT_NAME}.key"
+    CA_EXPIRE_DAYS="${CA_EXPIRE_DAYS:-365}"
+
+    LOCAL_CERT_DIR=/etc/tedge/device-certs/local-certs
+    LOCAL_CERT_NAME="${DEVICE_ID}-${SERVICE}"
+    LOCAL_CERT_FILE="${LOCAL_CERT_DIR}/${LOCAL_CERT_NAME}.crt"
+    LOCAL_CERT_KEY="${LOCAL_CERT_DIR}/${LOCAL_CERT_NAME}.key"
+    LOCAL_SIGNING_CERT_FILE="${LOCAL_CERT_DIR}/${SIGNING_CERT_NAME}.crt"
+
+    "${EXEC_CMD[@]}" "$TARGET" mkdir -p "$LOCAL_CERT_DIR"
+
+    if ! "${EXEC_CMD[@]}" "$TARGET" test -f "$LOCAL_CERT_KEY"; then
+        info "Creating device private key"
+        "${EXEC_CMD[@]}" "$TARGET" $SUDO openssl genrsa -out "$LOCAL_CERT_KEY" 2048
+        "${EXEC_CMD[@]}" "$TARGET" sudo chown tedge:root "$LOCAL_CERT_KEY"
+        "${EXEC_CMD[@]}" "$TARGET" sudo chmod 400 "$LOCAL_CERT_KEY"
+    else
+        info "Using existing private key"
+    fi
+
+    # Create Certificate Signing Request (CSR)
+    DEVICE_CSR=$(
+        "${EXEC_CMD[@]}" "$TARGET" $SUDO openssl req \
+            -key "$LOCAL_CERT_KEY" \
+            -new \
+            -subj "/O=thin-edge/OU=${DEVICE_ID}/CN=${SERVICE}"
+    )
+
+    # Sign the CSR
+    if [ "$SERVICE" = "child-agent" ]; then
+    CERT_EXT=$(cat << EOF
+basicConstraints=CA:FALSE
+extendedKeyUsage = clientAuth
+EOF
+    )
+    else
+    CERT_EXT=$(cat << EOF
+authorityKeyIdentifier=keyid
+basicConstraints=CA:FALSE
+keyUsage = digitalSignature, keyAgreement
+extendedKeyUsage = serverAuth, clientAuth
+subjectAltName = DNS:${TARGET_DNS},IP:${TARGET_IP},DNS:localhost,IP:127.0.0.1
```

Comment:
Is the IP address strictly required? I can imagine that having an IP address is potentially "more secure" however if the device changes ip address, then it require the users to reconfigure the certificates (which might not be feasible).

Reply:
The IP address is not required ... unless you have a local network without a DNS. Agree that this is fragile when combined with DHCP.

Reply:
We'll use IP addresses for now, but once https://github.com/thin-edge/thin-edge.io/issues/2803 is resolved we can revisit this decision.